### PR TITLE
Refactor/deprecate trace config

### PR DESCRIPTION
**Pull request by pitoniak32**
resolves: #2294 

## Changes
Moving away from `trace::Config` and `.with_config(...)` to remain consistent with the MeterProvider, and `LoggerProvider`.

# Migration Guide

Moving away from `TracerProvider::builder().with_config(trace::Config::default())`.

## Simple Example
The provider interface should be similar to the old method. For example, a `TracerProvider` being configured.
```rust
// old
let tracer_provider: TracerProvider = TracerProvider::builder()
    .with_config(Config::default().with_resource(Resource::empty()))
    .build();

// new
let tracer_provider: TracerProvider = TracerProvider::builder()
    .with_resource(Resource::empty())
    .build();
```

## Advanced Example
```rust
// old
let tracer_provider = TracerProvider::builder()
    .with_config(
        Config::default()
            .with_resource(Resource::empty())
            .with_sampler(Sampler::AlwaysOn)
            .with_max_events_per_span(16),
    )
    .build();

// new
let tracer_provider = TracerProvider::builder()
    .with_resource(Resource::empty())
    .with_sampler(Sampler::AlwaysOn)
    .with_max_events_per_span(16)
    .build();
```

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: examples/tracing-jaeger/src/main.rs
Comment (cijothomas) [inline examples/tracing-jaeger/src/main.rs:20]: we could potentially offer an easier ways for users who is just trying to add service name. For another time!

File: opentelemetry-sdk/src/trace/provider.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/trace/provider.rs:313]: Please add a changelog too.

**Review by cijothomas — APPROVED** [review]
LGTM. Lets add changelog too, will merge after one more approval.

Note: This will be part of 0.27.1 as there is no actual method removal, just deprecation warning this time. Actual removal will be 0.28.0 time only

File: opentelemetry-sdk/src/trace/provider.rs
Comment (lalitb) [inline opentelemetry-sdk/src/trace/provider.rs:333]:
```suggestion
    /// Specify the maximum number of events that can be recorded per span.
```

File: opentelemetry-sdk/src/trace/provider.rs
Comment (lalitb) [inline opentelemetry-sdk/src/trace/provider.rs:339]:
```suggestion
    /// Specify the maximum number of attributes that can be recorded per span.
```

File: opentelemetry-sdk/src/trace/provider.rs
Comment (lalitb) [inline opentelemetry-sdk/src/trace/provider.rs:345]:
```suggestion
    /// Specify the maximum number of links that can be recorded per span.
```

File: opentelemetry-sdk/src/trace/provider.rs
Comment (lalitb) [inline opentelemetry-sdk/src/trace/provider.rs:351]:
```suggestion
    /// Specify the maximum number of attributes one event can have.
```

File: opentelemetry-sdk/src/trace/provider.rs
Comment (lalitb) [inline opentelemetry-sdk/src/trace/provider.rs:357]:
```suggestion
    /// Specify the maximum number of attributes one link can have.
```

File: opentelemetry-zipkin/src/exporter/mod.rs
Comment (lalitb) [inline opentelemetry-zipkin/src/exporter/mod.rs:109]: Would be good to cleanup this instead of having deprecated code. Not a blocker for this PR though :)

**Review by lalitb — APPROVED** [review]
LGTM, with nit comments.

File: opentelemetry-zipkin/src/exporter/mod.rs
Comment (pitoniak32) [reply opentelemetry-zipkin/src/exporter/mod.rs:109]: it seems like Zipkin needs to be updated to follow the builder pattern that was added for otlp also, its currently using the pipeline version.

I will try and get this refactored to not use the allow.

Comment (pitoniak32) [reply opentelemetry-zipkin/src/exporter/mod.rs:109]: maybe there can be a followup for the zipkin to match the otlp crate.

Comment (cijothomas) [reply opentelemetry-zipkin/src/exporter/mod.rs]: Its best to do as a follow up to keep the scope for this PR smaller.

Comment (pitoniak32) [reply opentelemetry-zipkin/src/exporter/mod.rs]: yup, that's what I was thinking. I went simple with the approach on this PR, but I would like to open another to move to the Builder pattern if that would be welcomed

Comment (pitoniak32) [reply opentelemetry-zipkin/src/exporter/mod.rs]: my current change actually ignores the config, I'm going to revert to using the deprecated annotation, and then do a followup PR to move to the builder pattern to avoid too many changes in this PR.

**Pull request merged**
